proxy/tests/telemetry.rs

```diff
@@ -678,6 +678,223 @@ fn metrics_endpoint_outbound_request_duration() {
+                (String::from("addr_label2"), String::from("bar")),
+                (String::from("addr_label1"), String::from("foo")),
+            ],
+            Vec::new(),
+        );
+
+        info!("client.get(/)");
+        assert_eq!(client.get("/"), "hello");
+        // We can't make more specific assertions about the metrics
+        // besides asserting that both labels are present somewhere in the
+        // scrape, because testing for whole metric lines would depend on
+        // the order in which the labels occur, and we can't depend on hash
+        // map ordering.
+        assert_contains!(metrics.get("/metrics"), "dst_addr_label1=\"foo\"");
+        assert_contains!(metrics.get("/metrics"), "dst_addr_label2=\"bar\"");
+    }
+
+    #[test]
+    fn multiple_addrset_labels() {
+        let _ = env_logger::try_init();
+        let Fixture { client, metrics, proxy: _proxy } = fixture (
+            Vec::new(),
+            vec![
+                (String::from("set_label1"), String::from("foo")),
+                (String::from("set_label2"), String::from("bar")),
+            ]
+        );
+
+        info!("client.get(/)");
+        assert_eq!(client.get("/"), "hello");
+        // We can't make more specific assertions about the metrics
+        // besides asserting that both labels are present somewhere in the
+        // scrape, because testing for whole metric lines would depend on
+        // the order in which the labels occur, and we can't depend on hash
+        // map ordering.
+        assert_contains!(metrics.get("/metrics"), "dst_set_label1=\"foo\"");
+        assert_contains!(metrics.get("/metrics"), "dst_set_label2=\"bar\"");
+    }
+
+    #[test]
+    fn labeled_addr_and_addrset() {
+        let _ = env_logger::try_init();
+        let Fixture { client, metrics, proxy: _proxy } = fixture(
+            vec![(String::from("addr_label"), String::from("foo"))],
+            vec![(String::from("set_label"), String::from("bar"))],
+        );
+
+        info!("client.get(/)");
+        assert_eq!(client.get("/"), "hello");
+        assert_contains!(metrics.get("/metrics"),
+            "request_duration_ms_count{authority=\"labeled.test.svc.cluster.local\",direction=\"outbound\",dst_addr_label=\"foo\",dst_set_label=\"bar\"} 1");
+        assert_contains!(metrics.get("/metrics"),
+            "response_duration_ms_count{authority=\"labeled.test.svc.cluster.local\",direction=\"outbound\",dst_addr_label=\"foo\",dst_set_label=\"bar\",classification=\"success\",status_code=\"200\"} 1");
+        assert_contains!(metrics.get("/metrics"),
+            "request_total{authority=\"labeled.test.svc.cluster.local\",direction=\"outbound\",dst_addr_label=\"foo\",dst_set_label=\"bar\"} 1");
+        assert_contains!(metrics.get("/metrics"),
+            "response_total{authority=\"labeled.test.svc.cluster.local\",direction=\"outbound\",dst_addr_label=\"foo\",dst_set_label=\"bar\",classification=\"success\",status_code=\"200\"} 1");
```

Comment:
Again, I think that it would be clearer to have a less specific test here, like in the previous tests. Really what we care about is that the label from each the set and the addr are in the output. We don't care about the other aspects of the output. It's confusing for the reader of this test to suddenly see these irrelevant things being tested. Also, when those other aspects change (e.g. we rename "classification" to "class"), we shouldn't have to update *this* test.

That said, it's not the end of the world either.

Reply:
Yeah, that's true. My objective was to assert that the labels were added to all the metrics associated with that request, rather than just asserting that the label was present _somewhere_ in the exported metrics.

Once again, I'd like to eventually refactor these tests to do some limited parsing of the exported metrics so we can make assertions that individual metrics contain certain labels or values without also making assertions about the other labels that aren't relevant to that test.